### PR TITLE
Miniscript tapscript support (Pr 2 of 3)

### PR DESCRIPTION
**Pull request by sanket1729**
Adds support for MultiA script fragment. Builds on top of #255 .

Do not merge this until we have a rust-bitcoin release

### CODE REVIEW

File: src/interpreter/mod.rs
Comment (apoelstra) [inline src/interpreter/mod.rs]: I think, rather than returning an error here, we should return a closure that always returns `false`. But I'm not sure.

File: src/miniscript/satisfy.rs
Comment (apoelstra) [inline src/miniscript/satisfy.rs:1126]: In c05eb7c70d087269cc52c156b4ae4eadd2657133:

Ok to leave this comment here, but TBH I don't think there's any way we could eliminate this privacy leak. If the first signature is missing then the resulting witnesses will have a different distribution than if it's not ... which will become apparent over time.

Although, I guess, in practice users are unlikely to reuse keys in visible ways, so an adversary will have only a one-shot view of the signature distribution.

File: src/interpreter/mod.rs
Comment (JeremyRubin) [reply src/interpreter/mod.rs]: maybe you want a closure that always returns a Result<bool, Error>?

File: src/descriptor/sh.rs
Comment (sanket1729) [reply src/descriptor/sh.rs:221]: Agreed, it is much safer to do so. This does not rely on size check being the only error.

File: src/interpreter/mod.rs
Comment (sanket1729) [reply src/interpreter/mod.rs]: This should return an error if the input_idx is out of bounds.

On related note, I think we need a major redesign of this for taproot support anyways. This is currently hacked together that supports segwitv0 and legacy sighash. For taproot, we need entire sets of prevouts, it may be worth considering redesigning this entire API (yet) again :P

File: src/miniscript/astelem.rs
Comment (dr-orlovsky) [inline src/miniscript/astelem.rs:582]: I do believe this should be `match`; otherwise with future TapLeaf version-defined post-TapScripts we will not have a compiler error here

File: src/miniscript/satisfy.rs
Comment (dr-orlovsky) [inline src/miniscript/satisfy.rs:43]: Btw we will have this type in `rust-bitcoin`: https://github.com/rust-bitcoin/rust-bitcoin/pull/669

As for the name, probably it should have `ECDSA` part (not simply `EC`)

File: src/miniscript/satisfy.rs
Comment (dr-orlovsky) [inline src/miniscript/satisfy.rs:42]: `PartialOrd, Ord, Hash`

File: src/miniscript/satisfy.rs
Comment (dr-orlovsky) [inline src/miniscript/satisfy.rs:118]: `PartialOrd, Ord, Hash`

File: src/miniscript/satisfy.rs
Comment (dr-orlovsky) [inline src/miniscript/satisfy.rs:119]: Related: https://github.com/rust-bitcoin/rust-bitcoin/pull/671

File: src/miniscript/satisfy.rs
Comment (dr-orlovsky) [inline src/miniscript/satisfy.rs:567]: With more post-TapScript contexts in the future we will get a bug in this code and compiler will not warn/error. So I propose to use `match` here and not `if`

**Review by dr-orlovsky — COMMENTED** [review]
utACK mod some code style comments and proposal to re-use types which will be introduced in `rust-bitcoin` (once they got merged).

Also think that `Tap` context should be renamed to `TapScript`, since taproot may have more scripts in the future and `Tap` context requirements cover only leaf hash version `0xC0`, but not other future versions with their own contexts

File: src/miniscript/astelem.rs
Comment (sanket1729) [reply src/miniscript/astelem.rs:582]: There is a match check that frag_name is either `multi` or `multi_a`. I don't see an issue with doing an if-else here. 

```rust
("multi", n) | ("multi_a", n) => {
  				// code..
                if frag_name == "multi" {
                    pks.map(|pks| Terminal::Multi(k, pks))
                } else {
                    // must be multi_a
                    pks.map(|pks| Terminal::MultiA(k, pks))
                }
            }
```

File: src/miniscript/satisfy.rs
Comment (sanket1729) [reply src/miniscript/satisfy.rs:43]: Good catch

File: src/miniscript/satisfy.rs
Comment (sanket1729) [reply src/miniscript/satisfy.rs:567]: Agreed.

**Comment by sanket1729** [comment]
Pushed another branch. Some portions of this PR were merged in rust-bitcoin, so the review here should be (slightly) easy. Since this depends on a specific from rust-bitcoin, it is not possible to run integration tests on these because the dependencies operate on rust-bitcoin release. 

I think I should have considered this before merging the integration tests PR, but now we have to wait for a rust-bitcoincore-rpc release too.

File: src/miniscript/satisfy.rs
Comment (sanket1729) [reply src/miniscript/satisfy.rs]: I think the last commit should make this cleaner. Let me know if you want me to reorder the commits.

**Comment by sanket1729** [comment]
Rebased on ready for review.

**Review by apoelstra — APPROVED** [review]
ACK 42e4c50689ee96fc0a806f02c88834f24d78fd59

I'm a little unsure about the last commit - splitting `NoChecks`. In this PR, `NoChecksSchnorr` is not actually used and the interpreter continues to be ECDSA-only. It's not obvious how to taproot-enable the interpreter, so I trust you have a plan in mind that will make use of the new split `NoChecks`.

**Comment by sanket1729** [comment]
> In this PR, NoChecksSchnorr is not actually used and the interpreter continues to be ECDSA-only. It's not obvious how to taproot-enable the interpreter, so I trust you have a plan in mind that will make use of the new split NoChecks. 

I do have a plan for it. Our test-cases actually fail if we don't have this. We do a fancy interpreter check after we finalize a psbt to see that everything is good before we spend it. In that we need to serialize a script, We need some way to figure out `Miniscript<Pk, NoChecks>` should serialize key as 33 byte or 32 byte one or which signature algorithm to use for verification. All other contexts have this information implicitly in them because of associated consensus key type.

This was also done partly to address https://github.com/rust-bitcoin/rust-miniscript/pull/275#discussion_r748818958 in this PR.

**Comment by apoelstra** [comment]
Ah, I forgot about the PSBT sanity check. Makes sense!

**Pull request merged**
